### PR TITLE
bugfix: CLDSRV-144 Do not allow non-printable characters in object names

### PR DESCRIPTION
**Pull request by miniscruff**
Return an error instead of allowing these characters.

### CODE REVIEW

**Review by alexanderchan-scality — APPROVED** [review]

File: lib/api/initiateMultipartUpload.js [hunk 1]
```diff
@@ -45,6 +46,13 @@ function initiateMultipartUpload(authInfo, request, log, callback) {
     log.debug('processing request', { method: 'initiateMultipartUpload' });
     const bucketName = request.bucketName;
     const objectKey = request.objectKey;
+
+    if (hasNonPrintables(objectKey)) {
+        return callback(errors.InvalidInput.customizeDescription(
+            'object keys can not contain non-printable characters',
```
Comment (jonathan-gramain) [inline lib/api/initiateMultipartUpload.js]: Spelling suggestion

```suggestion
            'object keys cannot contain non-printable characters',
```

File: lib/utilities/stringChecks.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+ * hasNonPrintables returns whether or not the given value
+ * includes characters that are non-printable.
+ *
+ * @param {string} value - value to check for non-printables
+ * @returns {Boolean} whether or not the value has non-printables
+ */
+function hasNonPrintables(value) {
+    for (const char of value) {
+        const codePoint = char.codePointAt(0);
+        if (codePoint < 32) {
```
Comment (jonathan-gramain) [inline lib/utilities/stringChecks.js]: 127 is also nonprintable:

```suggestion
        if (codePoint < 32 || codePoint === 127) {
```

File: lib/api/objectPut.js [hunk 3]
```diff
@@ -47,6 +48,13 @@ function objectPut(authInfo, request, streamingV4Params, log, callback) {
     const requestType = 'objectPut';
     const valParams = { authInfo, bucketName, objectKey, requestType };
     const canonicalID = authInfo.getCanonicalID();
+
+    if (hasNonPrintables(objectKey)) {
+        return callback(errors.InvalidInput.customizeDescription(
+            'object keys can not contain non-printable characters',
```
Comment (jonathan-gramain) [inline lib/api/objectPut.js]: I thought you would change this one too ;)

```suggestion
            'object keys cannot contain non-printable characters',
```

**Review by jonathan-gramain — APPROVED** [review]

File: lib/api/objectPut.js (hunk 3 above)
Comment (miniscruff) [reply lib/api/objectPut.js]: oh haha, I just did the two suggestions sorry.

**Pull request merged**